modmail/extensions/extension_manager.py

```diff
@@ -0,0 +1,278 @@
+
+        If '\*' or '\*\*' is given as the name, all loaded extensions will be unloaded.
+        """  # noqa: W605
+        if not extensions:
+            await ctx.send_help(ctx.command)
+            return
+
+        blacklisted = [ext for ext in self.get_black_listed_extensions() if ext in extensions]
+
+        if blacklisted:
+            bl_msg = "\n".join(blacklisted)
+            await ctx.send(f":x: The following {self.type}(s) may not be unloaded:```\n{bl_msg}```")
+            return
+
+        if "*" in extensions or "**" in extensions:
+            extensions = sorted(
+                ext for ext in self.bot.extensions.keys() if ext not in self.get_black_listed_extensions()
+            )
+
+        await ctx.send(self.batch_manage(Action.UNLOAD, *extensions))
+
+    @extensions_group.command(name="reload", aliases=("r",))
+    async def reload_extensions(self, ctx: Context, *extensions: ExtensionConverter) -> None:
+        """
+        Reload extensions given their fully qualified or unqualified names.
+
+        If an extension fails to be reloaded, it will be rolled-back to the prior working state.
+
+        If '\*' is given as the name, all currently loaded extensions will be reloaded.
+        If '\*\*' is given as the name, all extensions, including unloaded ones, will be reloaded.
+        """  # noqa: W605
+        if not extensions:
+            await ctx.send_help(ctx.command)
+            return
+
+        if "**" in extensions:
+            extensions = self.all_extensions.keys()
+        elif "*" in extensions:
+            extensions = [*extensions, *sorted(self.bot.extensions.keys())]
+            extensions.remove("*")
+
+        await ctx.send(self.batch_manage(Action.RELOAD, *extensions))
+
+    @extensions_group.command(name="list", aliases=("all", "ls"))
+    async def list_extensions(self, ctx: Context) -> None:
+        """
+        Get a list of all extensions, including their loaded status.
+
+        Grey indicates that the extension is unloaded.
+        Green indicates that the extension is currently loaded.
+        """
+        embed = Embed(colour=Colour.blurple())
+        embed.set_author(
+            name=f"{self.type.capitalize()} List",
+        )
+
+        lines = []
+        categories = self.group_extension_statuses()
+        for category, extensions in sorted(categories.items()):
+            # Treat each category as a single line by concatenating everything.
+            # This ensures the paginator will not cut off a page in the middle of a category.
+            category = category.replace("_", " ").title()
+            extensions = "\n".join(sorted(extensions))
+            lines.append(f"**{category}**\n{extensions}\n")
+
+        log.debug(f"{ctx.author} requested a list of all {self.type}s. " "Returning a paginated list.")
+
+        # since we currently don't have a paginator.
+        await ctx.send("".join(lines) or f"( There are no {self.type}s installed. )")
```

Comment:
We should check if no types are installed before we start looping through them to form the message. It is more clear to what it is exactly doing and maybe even saves negligible performance.

Reply:
If lines is empty, it'll be a blank return and instantly go to the next. For the amount of times there *will* be something installed, this is faster overall.

Reply:
If we are going to check for installed before we create the message (which is reasonable) then it should be a guard statement at the top of the function.

Reply:
> If lines is empty, it'll be a blank return and instantly go to the next. For the amount of times there will be something installed, this is faster overall.

Hence, I said it as ineligible performance and focused more on clarity of the functionality, but that's my opinion and the reason bast said.